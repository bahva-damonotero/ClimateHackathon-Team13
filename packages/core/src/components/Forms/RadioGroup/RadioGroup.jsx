import React from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash.isequal';

import { flattenChildren } from '@boozallen-uip/uswds-react-utils';

import { Fieldset, FormError } from '..';

export default function RadioGroup({
  children,
  comparator,
  disabled,
  error,
  id,
  onChange,
  readOnly,
  tile,
  value
}) {
  let label;
  const options = [];

  flattenChildren(React.Children.toArray(children)).forEach((child) => {
    const componentType = child.type.displayName;

    switch (componentType) {
      case 'Legend':
        label = child;
        break;

      case 'Radio':
        options.push(child);
        break;

      default:
        throw new Error(
          `Invalid child component '${componentType}' used in RadioGroup.`
        );
    }
  });

  if (!label) {
    throw new Error('A Legend is required inside of a RadioGroup.');
  }

  if (!options.length) {
    throw new Error(
      'At least one Radio must be present inside of a RadioGroup.'
    );
  }

  return (
    <Fieldset>
      {React.cloneElement(label, { error })}

      <FormError error={error} />

      {options.map((child, index) => {
        const optionValue = child.props.value
          ? child.props.value
          : child.props.children;

        const checked = (() => {
          if (comparator) {
            return comparator(value, optionValue);
          }
          if (['function', 'object'].includes(typeof optionValue)) {
            return isEqual(value, optionValue);
          }
          return value === optionValue;
        })();

        const onChangeHandler = onChange
          ? (newValue) => {
              const event = new CustomEvent('change', {
                detail: {
                  name: id,
                  value: newValue
                }
              });
              onChange(event);
            }
          : undefined;

        return React.cloneElement(child, {
          checked,
          disabled: disabled || child.props.disabled,
          groupId: id,
          groupIndex: index,
          onChange: onChangeHandler,
          readOnly,
          tile
        });
      })}
    </Fieldset>
  );
}

RadioGroup.propTypes = {
  /** `<Legend>` + `<Radio>` node(s) */
  children: PropTypes.node.isRequired,
  /**
   * Let's you specify a custom function for comparing the selected value with the list of options.
   * If this is not provided, lodash's isEqual function is used: https://lodash.com/docs/#isEqual.
   */
  comparator: PropTypes.func,
  /** Should all the radio options be disabled? */
  disabled: PropTypes.bool,
  /** Error message to display */
  error: PropTypes.string,
  /** ID for the group of radio options */
  id: PropTypes.string.isRequired,
  /** Function called whenever the radio selection is changed */
  onChange: PropTypes.func,
  /** Should the radio options be marked as read-only? */
  readOnly: PropTypes.bool,
  /** If `true`, the tile radio variant is used for the entire group */
  tile: PropTypes.bool,
  /** Used to set the radio selection */
  value: PropTypes.any
};

RadioGroup.defaultProps = {
  disabled: false,
  readOnly: false,
  tile: false
};
