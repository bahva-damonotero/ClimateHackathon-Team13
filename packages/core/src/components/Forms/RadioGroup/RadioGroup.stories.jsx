/* eslint-disable react/prop-types */
import React from 'react';
import { useState } from '@storybook/addons';

import RadioGroup from '.';
import { Form, Legend, Radio, RadioLabelDescription } from '..';

export default {
  title: 'Forms/RadioGroup',
  component: RadioGroup,
  subcomponents: {
    Legend,
    Radio,
    RadioLabelDescription
  },
  argTypes: {
    label: {
      name: '<Legend> children',
      description: 'Label for the field',
      type: { required: true },
      table: {
        type: { summary: 'node' }
      }
    },
    screenReaderOnly: {
      name: '<Legend> screenReaderOnly',
      description:
        'If `true`, the label will only be visible to screen readers.',
      control: {
        type: 'boolean'
      },
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'false' }
      }
    },
    children: {
      control: { type: null }
    }
  }
};

const genders = [
  { key: 'M', value: 'Male' },
  { key: 'F', value: 'Female' },
  { key: 'O', value: 'Other' },
  { disabled: true, key: 'P', value: 'Prefer not to say' }
];

function renderOptions() {
  return genders.map((option) => (
    <Radio disabled={option.disabled} key={option.key} value={option}>
      {option.value}
    </Radio>
  ));
}

const DefaultTemplate = ({ label, screenReaderOnly, ...args }) => {
  const [value, setValue] = useState();

  return (
    <Form>
      <RadioGroup
        onChange={(e) => setValue(e.detail.value)}
        value={value}
        {...args}
      >
        <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

        {renderOptions()}
      </RadioGroup>
    </Form>
  );
};

const defaultArgs = {
  label: 'Gender',
  id: 'radio-group-example'
};

export const def = DefaultTemplate.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

const SelectionTemplate = ({ label, screenReaderOnly, ...args }) => (
  <Form>
    <RadioGroup {...args}>
      <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

      {renderOptions()}
    </RadioGroup>
  </Form>
);

export const selection = SelectionTemplate.bind({});
selection.args = {
  ...defaultArgs,
  readOnly: true,
  value: genders[1]
};

export const error = SelectionTemplate.bind({});
error.args = {
  ...defaultArgs,
  readOnly: true,
  error: 'Please choose an option'
};

export const disabled = DefaultTemplate.bind({});
disabled.args = {
  ...defaultArgs,
  disabled: true
};

export const customComparatorFunction = SelectionTemplate.bind({});
customComparatorFunction.args = {
  ...defaultArgs,
  comparator: (value, option) => value.key === option.key,
  readOnly: true,
  value: genders[2]
};

const SimpleUsageTemplate = ({ label, screenReaderOnly, ...args }) => (
  <Form>
    <RadioGroup {...args}>
      <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

      {genders.map((option) => (
        <Radio disabled={option.disabled} key={option.key}>
          {option.value}
        </Radio>
      ))}
    </RadioGroup>
  </Form>
);

export const simpleUsage = SimpleUsageTemplate.bind({});
simpleUsage.storyName = 'Simple Usage (no value props)';
simpleUsage.args = {
  ...defaultArgs,
  readOnly: true,
  value: genders[1].value
};

export const tile = DefaultTemplate.bind({});
tile.args = {
  ...defaultArgs,
  tile: true
};
const TileWithOptionsTextTemplate = ({ label, screenReaderOnly, ...args }) => {
  const [value, setValue] = useState();

  return (
    <Form>
      <RadioGroup
        onChange={(e) => setValue(e.detail.value)}
        value={value}
        {...args}
      >
        <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

        <Radio value="choiceOne" checked={value === 'choiceOne'}>
          Choice One
        </Radio>

        <Radio value="choiceTwo" checked={value === 'choiceTwo'}>
          Choice Two
          <RadioLabelDescription>
            This is optional text that can be used to describe the label in more
            detail.
          </RadioLabelDescription>
        </Radio>

        <Radio disabled>Choice Three</Radio>
      </RadioGroup>
    </Form>
  );
};

export const tileWithOptionsText = TileWithOptionsTextTemplate.bind({});
tileWithOptionsText.args = {
  label: 'Label',
  id: 'radio-group-example',
  tile: true
};
