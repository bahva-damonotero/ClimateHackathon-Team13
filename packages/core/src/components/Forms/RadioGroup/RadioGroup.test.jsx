import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import RadioGroup from '.';
import { Legend, Radio } from '..';

describe('RadioGroup', () => {
  const id = 'radio-group-example';
  const label = 'Gender';

  const genders = [
    { key: 'M', value: 'Male' },
    { key: 'F', value: 'Female' },
    { key: 'O', value: 'Other' },
    { disabled: true, key: 'P', value: 'Prefer not to say' }
  ];

  const renderOptions = () => {
    return genders.map((option) => (
      <Radio disabled={option.disabled} key={option.key} value={option}>
        {option.value}
      </Radio>
    ));
  };

  const jsx = (
    <RadioGroup id={id} readOnly>
      <Legend screenReaderOnly>{label}</Legend>

      {renderOptions()}
    </RadioGroup>
  );

  it('should render correctly', () => {
    const { container } = render(jsx);

    expect(container.querySelector('.usa-fieldset .usa-legend').innerHTML).toBe(
      label
    );
  });

  it('should throw an error if a Legend is not provided', () => {
    expect(() => {
      render(
        <RadioGroup id={id} readOnly>
          {renderOptions()}
        </RadioGroup>
      );
    }).toThrow('A Legend is required inside of a RadioGroup.');
  });

  it('should throw an error if no Radio components are provided', () => {
    expect(() => {
      render(
        <RadioGroup id={id} readOnly>
          <Legend screenReaderOnly>{label}</Legend>
        </RadioGroup>
      );
    }).toThrow('At least one Radio must be present inside of a RadioGroup.');
  });

  it('should throw an error when an invalid component is used in the RadioGroup', () => {
    expect(() => {
      render(
        <RadioGroup id={id} readOnly>
          <Legend screenReaderOnly>{label}</Legend>

          {renderOptions()}

          <span>Some content that doesn't belong here.</span>
        </RadioGroup>
      );
    }).toThrow("Invalid child component 'undefined' used in RadioGroup.");
  });

  it('should call `onChange` when the value changes', () => {
    const onChange = jest.fn();

    const { container } = render(
      <RadioGroup id={id} onChange={onChange}>
        <Legend screenReaderOnly>{label}</Legend>

        {renderOptions()}
      </RadioGroup>
    );

    fireEvent.click(container.querySelectorAll('.usa-radio__input')[0]);

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: genders[0]
        }
      })
    );
  });

  it('should handle falsey values based on given options', () => {
    const onChange = jest.fn();

    const options = [
      { key: 'true', value: true },
      { key: 'false', value: false },
      { key: 'string', value: 'string' },
      { key: 'undefined', value: undefined },
      { key: 'zero', value: 0 }
    ];

    const renderRadioOptions = () => {
      return options.map((option) => (
        <Radio disabled={option.disabled} key={option.key} value={option}>
          {option.value}
        </Radio>
      ));
    };

    const { container } = render(
      <RadioGroup id={id} onChange={onChange}>
        <Legend screenReaderOnly>{label}</Legend>

        {renderRadioOptions()}
      </RadioGroup>
    );

    fireEvent.click(container.querySelectorAll('.usa-radio__input')[0]);

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: options[0]
        }
      })
    );

    fireEvent.click(container.querySelectorAll('.usa-radio__input')[1]);

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: options[1]
        }
      })
    );

    fireEvent.click(container.querySelectorAll('.usa-radio__input')[3]);

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: options[3]
        }
      })
    );

    fireEvent.click(container.querySelectorAll('.usa-radio__input')[4]);

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: options[4]
        }
      })
    );
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
