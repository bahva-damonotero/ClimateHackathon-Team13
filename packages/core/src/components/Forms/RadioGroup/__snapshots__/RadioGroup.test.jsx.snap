// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should match snapshot 1`] = `
<fieldset
  className="usa-fieldset"
>
  <legend
    className="usa-legend usa-sr-only"
  >
    Gender
  </legend>
  <div
    className="usa-radio"
  >
    <input
      checked={false}
      className="usa-radio__input"
      disabled={false}
      id="radio-group-example--0"
      name="radio-group-example"
      readOnly={true}
      type="radio"
    />
    <label
      className="usa-radio__label"
      htmlFor="radio-group-example--0"
    >
      Male
    </label>
  </div>
  <div
    className="usa-radio"
  >
    <input
      checked={false}
      className="usa-radio__input"
      disabled={false}
      id="radio-group-example--1"
      name="radio-group-example"
      readOnly={true}
      type="radio"
    />
    <label
      className="usa-radio__label"
      htmlFor="radio-group-example--1"
    >
      Female
    </label>
  </div>
  <div
    className="usa-radio"
  >
    <input
      checked={false}
      className="usa-radio__input"
      disabled={false}
      id="radio-group-example--2"
      name="radio-group-example"
      readOnly={true}
      type="radio"
    />
    <label
      className="usa-radio__label"
      htmlFor="radio-group-example--2"
    >
      Other
    </label>
  </div>
  <div
    className="usa-radio"
  >
    <input
      checked={false}
      className="usa-radio__input"
      disabled={true}
      id="radio-group-example--3"
      name="radio-group-example"
      readOnly={true}
      type="radio"
    />
    <label
      className="usa-radio__label"
      htmlFor="radio-group-example--3"
    >
      Prefer not to say
    </label>
  </div>
</fieldset>
`;
